[Woo POS] Adjust pagination height threshold. Fix `GhostItemCardView` rendering when no more items.

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of https://github.com/woocommerce/woocommerce-ios/issues/14186
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
This PR addresses 2 items:
- Adds a threshold to infinite scrolling, so this is triggered when we've scrolled 70% through the current loaded view height
- Fixes an issue where the `GhostItemCardView` would always show at the end of the list when there are no more items to fetch from the remote.

https://github.com/user-attachments/assets/58608361-1289-4472-a1b8-d294bb294856


## Testing:
- For easiness, update the `ProductsRemote`'s `Constants.productsPerPage` to a lower threshold, like 7 or 10.
- Enter POS.
- Scroll down slowly, observe the right side of the list how as you scroll it re-positions itself as loads more items
- Scroll down fast, observe that if you reach the end of the currently-loaded item list, the temporary `GhostItemCardView` will appear while loading data
- Scroll to the end of the list, the `GhostItemCardView` may appear once, but no more after that.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Ad threshold multipler to scrolling

iamgabrielma committed:
Avoid infinite loop of loading GhostItemCardView

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -207,6 +207,8 @@ private extension ItemListView {
         static let itemListPadding: CGFloat = 16
         static let bannerCardPadding: CGFloat = 16
+        static let viewHeight: CGFloat = UIScreen.main.bounds.height
+        static let scrollThresholdMultiplier: CGFloat = 1.7
```
To clarify, does this multiplier mean the list is scrolled 70% of its total height?

iamgabrielma replied:
Yes, that's right!

staskus reviewed (commented):
Thanks for the work!

I'm not sure if it's connected to this change but something in this behavior breaks after doing pull to refresh:

1. Pull to refresh triggers full screen reload, rather than reload of table
2. Ghost cells are no longer shown

https://github.com/user-attachments/assets/8e233b98-6921-4567-83ff-e8190f0ec145

iamgabrielma committed:
Merge branch 'trunk' into task/pos-pagination-adjust-threshold

iamgabrielma committed:
Do not render ghost card on 1st pull-to-refresh

iamgabrielma committed:
linting

iamgabrielma commented:
Thanks for the review! 

> Pull to refresh triggers full screen reload, rather than reload of table

I just merged trunk and this should be resolved now 😅 

> Ghost cells are no longer shown

That's an interesting case, these will render perfectly fine if we just scroll down, but will stop rendering if we pull-to-refresh, and then scroll down. 

This seem to happen because when we pull-to-refresh, we load the initial items, and then fetch the next page, but `uniqueNewItems.count` at that point are zero, so it sets `hasMoreItems = false` which stops rendering the GhostItemCardView. 

I spent a couple of hours but I seem unable to find a way through that works 100% of the times. When we reach the end of the page, since we do not know if there are more pages, if we pull-to-refresh we either break the GhostItemCardView rendering, or we break the loading state and triggers a full screen reload, or we have to keep the "infinite" loading at the end of the last item🤔  I'm a bit at a lost here, so any advice is appreciated @staskus

iamgabrielma commented:
> I spent a couple of hours but I seem unable to find a way through that works 100% of the times. When we reach the end of the page, since we do not know if there are more pages, if we pull-to-refresh we either break the GhostItemCardView rendering, or we break the loading state and triggers a full screen reload, or we have to keep the "infinite" loading at the end of the last item🤔 I'm a bit at a lost here, so any advice is appreciated @staskus

Adding to this, something I will try tomorrow is to compare if the last item fetched on n-1 is same as last item fetched on n, if that's the case we could trigger something like "isLastPage" which would resolve the infinite "loading" loop at the end of scrolling without having to rely on view state. The only issue that I foresee is that it wouldn't update new products added via web by scrolling down, and the merchant would have to pull-to-refresh to see these, so maybe is a compromise we have to take for now.

iamgabrielma committed:
Assume there’s always more items on reload

We do not need to remove the items since we’re re-setting the list entirely on reload

iamgabrielma commented:
Alright, this took longer than expected but seems to be fixed with one liner by updating state assuming that there always be more items if we reload:

![Simulator Screen Recording - iPad Air 11 - iOS 17 5 M2 - 2024-10-31 at 09 13 10](https://github.com/user-attachments/assets/65c4f42a-4aa7-4a46-8b08-6eddd771e90c)

- When there are more items, ghost card is rendered
- When reaching the bottom of the view, ghost card is hidden
- Pull-to-refresh doesn't break behaviour

On a related note, I think we still need to make some improvements on how we handle view state around here that should make future changes more bulletproof

staskus commented:
Thanks for the fixes!

> I spent a couple of hours but I seem unable to find a way through that works 100% of the times. 

Yes, ideally there should be a stronger relationship with the API that tells us the total number of pages or products. Same with pull to refresh, to understand if there are any updates - if not don't refresh the existing list. Otherwise we need to play around with the view state.

> Alright, this took longer than expected but seems to be fixed with one liner by updating state assuming that there always be more items if we reload

Unfortunately, I'm able to reproduce a similar issue 🤔 

When we do pull to refresh, sometimes two identical requests happen at the same time. The first one finishes with a list of new items, and quickly after that a second request finishes. However, since it was the same request there were no unique items, and `hasMoreItems` is set to `false`.

It happens when `productsPerPage` is low (7), I think at the same time we make pull to refresh also the threshold is triggered. But it can be reproduced with a higher threshold if you scroll down immediately after scrolling down. 

https://github.com/user-attachments/assets/cc94c0c7-d0e2-45b9-9fc5-d8f9147ef86d

---------------
A few other things I noticed:
1. We keep trying to fetch new items even if we don't show ghost cells anymore, intentional?
2. We try to request a new cell not only when we go down but also when we scroll up. It's partially related to point number 1 since it only happens when the full list is loaded.

iamgabrielma committed:
Disallow loading calls when scrolling up

We save and perform an additional check against the lastScrollPosition before fetching the next page, this avoids unnecessary network calls when we’re scrolling up through already loaded products.

iamgabrielma committed:
Disallow refresh if it’s already loading

If we pull-to-refresh fast enough, the system might attempt to reload while the state is still loading

iamgabrielma commented:
Thanks again for you review Povilas! 

> We try to request a new cell not only when we go down but also when we scroll up. It's partially related to point number 1 since it only happens when the full list is loaded.

Good catch, I added an additional check on https://github.com/woocommerce/woocommerce-ios/pull/14234/commits/117ead603b9ebbbf679096aa41aa1eba36860396 so we're sure to not perform additional calls when scrolling back up.

> When we do pull to refresh, sometimes two identical requests happen at the same time. The first one finishes with a list of new items, and quickly after that a second request finishes. However, since it was the same request there were no unique items, and hasMoreItems is set to false. It happens when productsPerPage is low (7), I think at the same time we make pull to refresh also the threshold is triggered. But it can be reproduced with a higher threshold if you scroll down immediately after scrolling down.

This one is tricky, something I added is an additional check to disallow PTR if state is still loading here: https://github.com/woocommerce/woocommerce-ios/pull/14234/commits/b52930167c54192535de64609a97b966a88324e8, which was possible up until now if we PTR fast enough. But yes, I can see this happening when the list is short enough, since will load the initial items, and also attempt to load the next page since what's in screen triggers the threshold. And as you say, if there are no "new items" coming back, then `hasMoreItems` is set to false and remains like that when we scroll down.

https://github.com/user-attachments/assets/d432aa9f-7a21-4c43-b1dd-eadc2a97f129

I think here we need to make one of the compromises we talked about in the project thread: Since the API cannot help us, and as each network call should come back with 100 products (or a good part of them), having a low number of items on screen and this causing the ghost cell not rendering would be a an edge cases and can iterate further as we go. What do you think?

iamgabrielma commented:
I'm closing this one in favour of https://github.com/woocommerce/woocommerce-ios/pull/14382
